### ISSUE TITLE
panic: MaybeRestoreResourceInstanceDeposed called without DeposedKey

### ISSUE THREAD

**Issue by dekimsey** [hashicorp/terraform#22974]
### Terraform Version

```
Terraform v0.12.9
+ provider.aws v2.28.1
+ provider.azuread v0.6.0
+ provider.null v2.1.2
+ provider.random v2.2.1
+ provider.template v2.1.2
+ provider.tls v2.1.1
```

### Terraform Configuration Files

```hcl
# Our external zone for management and entry to the VPC.
resource "aws_route53_zone" "external" {
  name          = local.external-domain
  comment       = "Externally-facing, managed entirely by Terraform"
  force_destroy = true

  tags = {
    namespace = local.unique_namespace
    terraform = "true"
  }
}

resource "null_resource" "akamai-external-delegation" {
  # the akamai provider is dumb.
  # it conflates records & zones: https://github.com/akamai/terraform-provider-akamai/issues/56#issuecomment-498322386
  triggers = {
    name_servers = join(" ", aws_route53_zone.external.name_servers)
    domain       = local.external-domain
  }

  provisioner "local-exec" {
    command = "python scripts/update-akamai-zone --ns=${join(",", aws_route53_zone.external.name_servers)} trustwave.cloud ${replace(
      local.external-domain,
      format(".%s", local.external-parent-domain),
      "",
    )}"
  }
  provisioner "local-exec" {
    when = destroy
    command = "python scripts/update-akamai-zone --destroy trustwave.cloud ${replace(
      local.external-domain,
      format(".%s", local.external-parent-domain),
      "",
    )}"
  }
}
```

### Debug Output
```
2019/10/01 14:02:30 [DEBUG] null_resource.akamai-external-delegation: applying the planned Create change
2019/10/01 14:02:30 [TRACE] GRPCProvider: ApplyResourceChange
2019/10/01 14:02:30 [TRACE] <root>: eval: *terraform.EvalMaybeTainted
2019/10/01 14:02:30 [TRACE] <root>: eval: *terraform.EvalWriteState
2019/10/01 14:02:30 [TRACE] EvalWriteState: writing current state object for null_resource.akamai-external-delegation
2019/10/01 14:02:30 [TRACE] <root>: eval: *terraform.EvalApplyProvisioners
2019/10/01 14:02:30 [TRACE] EvalApplyProvisioners: provisioning null_resource.akamai-external-delegation with "local-exec"
2019/10/01 14:02:30 [TRACE] GetResourceInstance: aws_route53_zone.external is a single instance
2019/10/01 14:02:31 [WARN] Errors while provisioning null_resource.akamai-external-delegation with "local-exec", so aborting
2019/10/01 14:02:31 [TRACE] EvalApplyProvisioners: null_resource.akamai-external-delegation provisioning failed, but we will continue anyway at the caller's request
2019/10/01 14:02:31 [TRACE] <root>: eval: *terraform.EvalMaybeTainted
2019/10/01 14:02:31 [TRACE] EvalMaybeTainted: null_resource.akamai-external-delegation encountered an error during creation, so it is now marked as tainted
2019/10/01 14:02:31 [TRACE] <root>: eval: *terraform.EvalWriteState
2019/10/01 14:02:31 [TRACE] EvalWriteState: writing current state object for null_resource.akamai-external-delegation
2019/10/01 14:02:31 [TRACE] <root>: eval: *terraform.EvalIf
2019/10/01 14:02:31 [TRACE] <root>: eval: *terraform.EvalMaybeRestoreDeposedObject
2019/10/01 14:02:31 [TRACE] vertex "null_resource.akamai-external-delegation": visit complete
panic: MaybeRestoreResourceInstanceDeposed called without DeposedKey

goroutine 898 [running]:
github.com/hashicorp/terraform/states.(*SyncState).MaybeRestoreResourceInstanceDeposed(0xc002cd78c0, 0x0, 0x0, 0x0, 0x4d, 0xc000664ad0, 0xd, 0xc0002b0c00, 0x1a, 0x0, ...)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/states/sync.go:430 +0x20e
github.com/hashicorp/terraform/terraform.(*EvalMaybeRestoreDeposedObject).Eval(0xc0038f0e00, 0x2dccf60, 0xc0016611e0, 0x2, 0x2, 0x0, 0xc002f2c100)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval_state.go:391 +0x1ad
github.com/hashicorp/terraform/terraform.EvalRaw(0x2d5cb00, 0xc0038f0e00, 0x2dccf60, 0xc0016611e0, 0x0, 0x0, 0x0, 0x27)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalIf).Eval(0xc00385ea80, 0x2dccf60, 0xc0016611e0, 0x2, 0x2, 0x0, 0x0)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval_if.go:18 +0x114
github.com/hashicorp/terraform/terraform.EvalRaw(0x2d5ca40, 0xc00385ea80, 0x2dccf60, 0xc0016611e0, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc0033bb8e0, 0x2dccf60, 0xc0016611e0, 0x2, 0x2, 0x1aa5a45, 0x2d5cd00)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:20 +0xfd
github.com/hashicorp/terraform/terraform.EvalRaw(0x2d5ccc0, 0xc0033bb8e0, 0x2dccf60, 0xc0016611e0, 0x261c3a0, 0x3b5c3c1, 0x2588940, 0xc004360c90)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.Eval(0x2d5ccc0, 0xc0033bb8e0, 0x2dccf60, 0xc0016611e0, 0xc0033bb8e0, 0x2d5ccc0, 0xc0033bb8e0, 0x0)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/eval.go:35 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x29024a0, 0xc005652870, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/terraform/graph.go:90 +0xf7e
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc002e2c600, 0x29024a0, 0xc005652870, 0xc002805c00)
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x377
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/private/tmp/terraform-20190917-90047-18ajf1n/terraform-0.12.9/src/github.com/hashicorp/terraform/dag/walk.go:314 +0xaa7
```

### Crash Output

(Encrypted. Though there is no actual sensitive data, metadata is always useful.)
https://gist.github.com/dekimsey/fde76373218dfd567667142464ee15f5

### Expected Behavior
Run completed successfully.

### Actual Behavior
Crashed. Re-ran, no crash.

### Steps to Reproduce
terraform apply -target=null_resource.akamai-external-delegation

### Additional Context
We were troubleshooting (running multiple times) an issue with our local_provider script that was causing the script to stacktrace/fail. When we ran the apply the diff showed two items, one deposed and one to be created.

### References

None.

**Comment by apparentlymart** [comment]
I've done some further digging into this today and wanted to capture what I've learned, even though I don't have a concrete action plan just yet.

It seems like the root problem here is related to the condition here which is supposed to be guarding against that case:

https://github.com/hashicorp/terraform/blob/f1237f816c5d00c59648d3170c7b8f4afe6d982d/terraform/node_resource_apply_instance.go#L392-L400

This `deposedKey` variable supposed to get populated further up when we deposed the object in the first place:

https://github.com/hashicorp/terraform/blob/f1237f816c5d00c59648d3170c7b8f4afe6d982d/terraform/node_resource_apply_instance.go#L250-L266

We can see here that the guard conditions on both of these are different, and so it seems likely that the way we're getting to this panic is that the `EvalDeposedState` condition is returning `false` while the `EvalMaybeRestoreDeposedObject` condition is returning `true`, but it's not clear what would cause that to be true since the `EvalMaybeRestoreDeposedObject` condition uses `createBeforeDestroyEnabled` as part of its computation.

Unfortunately since this is a transient sort of issue that requires quite a specific sequence of events and tends to "self-heal" for a subsequent run :roll_eyes: I wasn't able to find a repeatable reproduction case for this in order to diagnose further what's causing the disconnect between these two.

For that reason, we're going to put in place an interim solution of detecting this situation and returning a normal error for it rather than panicking, and we'll include in that error message some additional context that would hopefully give us more information to inform a future reproduction and fix.

**Comment by danieldreier** [comment]
@apparentlymart I think that this was fixed in #23718 but not resolved here - is that accurate? Can we close this?

**Comment by apparentlymart** [comment]
The intent of leaving this open was to represent that we didn't actually fix the bug, and instead we just made it return an error instead of panicking when the panic occurs.

However, as far as I know nobody has reported seeing that error message (which explicitly asks to report it) since we released that change, which suggests that whatever caused this error to occur was so rare that I'm inclined to now close this issue and see if we get any new issues in future reporting it.

---

If you've found this issue in future because you encountered an error like the following, please _open a new issue_ to report it and mention this one (say `#22974` in your issue text) to create a link back here.

```
Error: Attempt to restore non-existent deposed object

Terraform has encountered a bug where it would need
to restore a deposed object for aws_instance.foo without
knowing a deposed object key for that object. This is a
bug in Terraform; please report it!
```
